task/power/src/main.rs

```diff
@@ -39,117 +39,100 @@ include!(concat!(env!("OUT_DIR"), "/i2c_config.rs"));
-    }
-}
-
-fn read_voltage<E, T: VoltageSensor<E>>(
-    device: &T,
-) -> Result<Volts, ResponseCode>
-where
-    ResponseCode: From<E>,
-{
-    match device.read_vout() {
-        Ok(reading) => Ok(reading),
-        Err(err) => {
-            let err: ResponseCode = err.into();
-            Err(err)
-        }
-    }
+enum Device {
+    Bmr491(Bmr491),
+    Raa229618(Raa229618),
+    Isl68224(Isl68224),
+    Tps546B24A(Tps546B24A),
+    Adm1272(Adm1272),
+    Max5970(Max5970),
 }
 
-impl PowerController {
+impl Device {
     fn read_temperature(&self) -> Result<Celsius, ResponseCode> {
-        match &self.device {
-            Device::IBC(dev) => read_temperature(dev),
-            Device::Core(dev) | Device::Mem(dev) => read_temperature(dev),
-            Device::SerDes(dev) => read_temperature(dev),
-            Device::Sys(dev) => read_temperature(dev),
-            Device::HotSwap(dev) | Device::Fan(dev) => read_temperature(dev),
-            _ => panic!(),
-        }
+        let r = match &self {
+            Device::Bmr491(dev) => dev.read_temperature()?,
+            Device::Raa229618(dev) => dev.read_temperature()?,
+            Device::Isl68224(dev) => dev.read_temperature()?,
+            Device::Tps546B24A(dev) => dev.read_temperature()?,
+            Device::Adm1272(dev) => dev.read_temperature()?,
+            Device::Max5970(_dev) => return Err(ResponseCode::NoDevice),
+        };
+        Ok(r)
     }
 
     fn read_iout(&self) -> Result<Amperes, ResponseCode> {
-        match &self.device {
-            Device::IBC(dev) => read_current(dev),
-            Device::Core(dev) | Device::Mem(dev) => read_current(dev),
-            Device::MemVpp(dev) => read_current(dev),
-            Device::SerDes(dev) => read_current(dev),
-            Device::Sys(dev) => read_current(dev),
-            Device::HotSwap(dev) | Device::Fan(dev) => read_current(dev),
-            Device::HotSwapIO(dev) => read_current(dev),
-        }
+        let r = match &self {
+            Device::Bmr491(dev) => dev.read_iout()?,
+            Device::Raa229618(dev) => dev.read_iout()?,
+            Device::Isl68224(dev) => dev.read_iout()?,
+            Device::Tps546B24A(dev) => dev.read_iout()?,
+            Device::Adm1272(dev) => dev.read_iout()?,
+            Device::Max5970(dev) => dev.read_iout()?,
+        };
+        Ok(r)
     }
 
     fn read_vout(&self) -> Result<Volts, ResponseCode> {
+        let r = match &self {
+            Device::Bmr491(dev) => dev.read_vout()?,
+            Device::Raa229618(dev) => dev.read_vout()?,
+            Device::Isl68224(dev) => dev.read_vout()?,
+            Device::Tps546B24A(dev) => dev.read_vout()?,
+            Device::Adm1272(dev) => dev.read_vout()?,
+            Device::Max5970(dev) => dev.read_vout()?,
+        };
+        Ok(r)
+    }
+}
+
+impl PowerControllerConfig {
+    fn get_device(&self, task: TaskId) -> Device {
+        let (dev, rail) = (self.builder)(task);
+        use DeviceType::*;
```

Comment:
Style nit - I'm not a fan of importing enum variants like this; match arms with just a single identifier open the door for typos to bind new variable names (e.g., https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7d0461a815e49ffbfe83c0f6a9b48bea). Usually (maybe always?) this produces warnings, but it will still compile.

Reply:
Okay, I'm persuaded and will use full names here